homeassistant/components/picnic/services.py

```diff
@@ -0,0 +1,131 @@
+"""Services for the Picnic integration."""
+from __future__ import annotations
+
+from typing import Any
+
+from python_picnic_api import PicnicAPI
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.helpers import device_registry
+import homeassistant.helpers.config_validation as cv
+
+from .const import (
+    ATTR_AMOUNT,
+    ATTR_DEVICE_ID,
+    ATTR_PRODUCT_ID,
+    ATTR_PRODUCT_IDENTIFIERS,
+    ATTR_PRODUCT_NAME,
+    CONF_API,
+    DOMAIN,
+    SERVICE_ADD_PRODUCT_TO_CART,
+)
+
+
+class PicnicServiceException(Exception):
+    """Exception for Picnic services."""
+
+
+async def async_register_services(hass: HomeAssistant) -> None:
+    """Register services for the Picnic integration, if not registered yet."""
+
+    # Register the add_product service
+    if not hass.services.has_service(DOMAIN, SERVICE_ADD_PRODUCT_TO_CART):
+
+        async def async_add_product_service(call: ServiceCall):
+            # Get the picnic API client and call the handler with api client and hass
+            api_client: PicnicAPI = await get_api_client(
+                hass, call.data.get("device_id")
+            )
+            await handle_add_product(hass, api_client, call)
+
+        hass.services.async_register(
+            DOMAIN,
+            SERVICE_ADD_PRODUCT_TO_CART,
+            async_add_product_service,
+            schema=vol.Schema(
+                {
+                    vol.Optional(ATTR_DEVICE_ID): cv.string,
+                    vol.Exclusive(
+                        ATTR_PRODUCT_ID, ATTR_PRODUCT_IDENTIFIERS
+                    ): cv.positive_int,
+                    vol.Exclusive(
+                        ATTR_PRODUCT_NAME, ATTR_PRODUCT_IDENTIFIERS
+                    ): cv.string,
+                    vol.Optional(ATTR_AMOUNT): vol.All(
+                        vol.Coerce(int), vol.Range(min=1)
+                    ),
+                }
+            ),
+        )
+
+
+async def get_api_client(hass: HomeAssistant, device_id: Any | None = None):
+    """Get the right Picnic API client based on the device id, else get the default one."""
+    if device_id is None:
+        default_config_id = list(hass.data[DOMAIN].keys())[0]
```

Comment:
I'm not sure if this is a pattern there is prior art for or not -- is this common to pick an arbitrary device in a service?

Reply:
Most people will have only one time the integration added as it's address based. So picking the first configured integration will work for 99% of the cases and it saves the user from configuring the device id for each service call.

Reply:
Oh let me clarify -- I wasn't asking *why*, as it's clear the motivation -- My question is if this is the correct design standard (e.g. potential for confusion/bugs/etc) and if this should be handled at a higher level similar to how entity matches are handled: https://github.com/home-assistant/core/blob/c741d9d0452970c39397deca1c65766c8cb917da/homeassistant/const.py#L64

Reply:
I can't find any references to handling this on a higher level. So the solution would be to make the `device_id` required?

Reply:
Yeah, that seems the most conservative approach for an API to me. Perhaps I suggest require device id and we'll get this PR merged -- then if you feel that this is needed, it can come in a smaller separate PR that can get a closer look from other core devs.  (Just a suggestion to get this merged faster without blocking on this smaller detail)

Reply:
Ok, then I'll make it required!